### PR TITLE
Eperez backdoors ng

### PR DESCRIPTION
**Pull request by enriquepablo**
Add some backdoor views to get verification codes in staging when sending a magic cookie. Views added:

* Get verification code for a new email; 
* Get verification code for a new phone number; 
* Get the emailed code for resetting the password;
* Get the SMSed code for resetting the password;
* Get verification code sent by email during signup;
* Also, sending a magic cookie will bypass recaptcha verification.
* For NIN veriification:
  * For eidas, just sending the cookie should verify the NIN. I am adding a function nin_verify_BACKDOOR to mock what real verification would do.
  * For letter proofing, there is a view to get the code sending a magic cookie.
  * for mobile proofing, again just sending the cookie should verify the NIN.

### CODE REVIEW

File: src/eduid_webapp/email/views.py [hunk 1]
```diff
@@ -292,4 +293,22 @@ def resend_code(user, email):
...
+    Backdoor to get the verification code in the staging or dev environments
+    """
+    try:
+        if check_magic_cookie(current_app.config):
+            eppn = request.args.get('eppn')
+            email = request.args.get('email')
+            state = current_app.proofing_statedb.get_state_by_eppn_and_email(eppn, email)
+            return state.verification.verification_code
+    except Exception:
+        pass
```
Comment (fredrikt) [inline src/eduid_webapp/email/views.py:312]: not sure what exceptions are caught here, but maybe log at least?

File: src/eduid_webapp/letter_proofing/views.py [hunk 2]
```diff
@@ -141,3 +150,19 @@ def verify_code(user, code):
...
+def get_code(user):
+    """
+    Backdoor to get the verification code in the staging or dev environments
+    """
+    try:
+        if check_magic_cookie(current_app.config):
+            state = current_app.proofing_statedb.get_state_by_eppn(user.eppn)
+            return state.nin.verification_code
+    except Exception:
+        pass
```
Comment (fredrikt) [inline src/eduid_webapp/letter_proofing/views.py:166]: logging here too?

File: src/eduid_webapp/lookup_mobile_proofing/helpers.py [hunk 3]
```diff
@@ -62,6 +63,27 @@ def match_mobile_to_user(user, self_asserted_nin, verified_mobile_numbers):
     :rtype: tuple
     """
     proofing_user = ProofingUser.from_user(user, current_app.private_userdb)
+
+    # This code is to use the backdoor that allows selenium integration tests
+    # to verify a NIN by sending a magic cookie
+    if check_magic_cookie(current_app.config):
+        current_app.logger.debug('Using the BACKDOOR to verify a NIN through the lookup mobile app')
```
Comment (fredrikt) [inline src/eduid_webapp/lookup_mobile_proofing/helpers.py:70]: info log this please

File: src/eduid_webapp/phone/views.py [hunk 4]
```diff
@@ -255,4 +255,21 @@ def resend_code(user, number):
     current_app.stats.count(name='mobile_resend_code', value=1)
 
     phones = {'phones': user.phone_numbers.to_list_of_dicts(), 'message': 'phones.code-sent'}
-    return phones
+    return PhoneListPayload().dump(phones)
```
Comment (fredrikt) [inline src/eduid_webapp/phone/views.py:258]: same here, MarshalWith should handle that

File: src/eduid_webapp/reset_password/views/reset_password.py [hunk 5]
```diff
@@ -517,3 +518,35 @@ def set_new_pw_extra_security_token() -> dict:
...
+    """
+    Backdoor to get the email verification code in the staging or dev environments
+    """
+    try:
+        if check_magic_cookie(current_app.config):
+            eppn = request.args.get('eppn')
+            state = current_app.password_reset_state_db.get_state_by_eppn(eppn)
+            return state.email_code.code
+    except Exception:
+        pass
```
Comment (fredrikt) [inline src/eduid_webapp/reset_password/views/reset_password.py:534]: log?

File: src/eduid_webapp/reset_password/views/reset_password.py [hunk 6]
```diff
@@ -517,3 +518,35 @@ def set_new_pw_extra_security_token() -> dict:
...
+    """
+    Backdoor to get the phone verification code in the staging or dev environments
+    """
+    try:
+        if check_magic_cookie(current_app.config):
+            eppn = request.args.get('eppn')
+            state = current_app.password_reset_state_db.get_state_by_eppn(eppn)
+            return state.phone_code.code
+    except Exception:
+        pass
```
Comment (fredrikt) [inline src/eduid_webapp/reset_password/views/reset_password.py:550]: log?

File: src/eduid_webapp/security/views/reset_password.py [hunk 7]
```diff
@@ -323,3 +324,35 @@ def new_password(state):
...
+    """
+    Backdoor to get the email verification code in the staging or dev environments
+    """
+    try:
+        if check_magic_cookie(current_app.config):
+            eppn = request.args.get('eppn')
+            state = current_app.password_reset_state_db.get_state_by_eppn(eppn)
+            return state.email_code.code
+    except Exception:
+        pass
```
Comment (fredrikt) [inline src/eduid_webapp/security/views/reset_password.py:340]: log?

File: src/eduid_webapp/security/views/reset_password.py [hunk 8]
```diff
@@ -323,3 +324,35 @@ def new_password(state):
...
+    """
+    Backdoor to get the phone verification code in the staging or dev environments
+    """
+    try:
+        if check_magic_cookie(current_app.config):
+            eppn = request.args.get('eppn')
+            state = current_app.password_reset_state_db.get_state_by_eppn(eppn)
+            return state.phone_code.code
+    except Exception:
+        pass
```
Comment (fredrikt) [inline src/eduid_webapp/security/views/reset_password.py:356]: log?

File: src/eduid_webapp/signup/views.py [hunk 9]
```diff
@@ -118,3 +126,20 @@ def verify_link(code):
...
+    Backdoor to get the email verification code in the staging or dev environments
+    """
+    try:
+        if check_magic_cookie(current_app.config):
+            email = request.args.get('email')
+            signup_user = current_app.private_userdb.get_user_by_pending_mail_address(email)
+            code = signup_user.pending_mail_address.verification_code
+            return code
+    except Exception:
+        pass
```
Comment (fredrikt) [inline src/eduid_webapp/signup/views.py:143]: log?

File: src/eduid_webapp/email/views.py [hunk 10]
```diff
@@ -292,4 +293,22 @@ def resend_code(user, email):
     current_app.stats.count(name='email_resend_code', value=1)
 
     emails = {'emails': user.mail_addresses.to_list_of_dicts(), 'message': 'emails.code-sent'}
-    return emails
+    return EmailListPayload().dump(emails)
```
[the remark this answers is not in the document]
Comment (enriquepablo) [reply src/eduid_webapp/email/views.py:296]: yes

File: src/eduid_webapp/email/views.py (hunk 1 above)
Comment (enriquepablo) [reply src/eduid_webapp/email/views.py:312]: Addressed - replying just to keep track :)

File: src/eduid_webapp/letter_proofing/views.py (hunk 2 above)
Comment (enriquepablo) [reply src/eduid_webapp/letter_proofing/views.py:166]: ack

File: src/eduid_webapp/reset_password/views/reset_password.py (hunk 5 above)
Comment (enriquepablo) [reply src/eduid_webapp/reset_password/views/reset_password.py:534]: ack

File: src/eduid_webapp/reset_password/views/reset_password.py (hunk 6 above)
Comment (enriquepablo) [reply src/eduid_webapp/reset_password/views/reset_password.py:550]: ack

File: src/eduid_webapp/security/views/reset_password.py (hunk 7 above)
Comment (enriquepablo) [reply src/eduid_webapp/security/views/reset_password.py:340]: ack

File: src/eduid_webapp/security/views/reset_password.py (hunk 8 above)
Comment (enriquepablo) [reply src/eduid_webapp/security/views/reset_password.py:356]: ack

File: src/eduid_webapp/phone/views.py (hunk 4 above)
Comment (enriquepablo) [reply src/eduid_webapp/phone/views.py:258]: Don't know what I was thinking...

Comment (enriquepablo) [reply src/eduid_webapp/phone/views.py:258]: Aw right, It was merging the upgrade of marshmallow

File: src/eduid_webapp/signup/views.py (hunk 9 above)
Comment (enriquepablo) [reply src/eduid_webapp/signup/views.py:143]: done

**Pull request merged**